### PR TITLE
feat(eks): option to disable manifest validation

### LINKED ISSUE #11763: (aws-eks): Add validate flag to KubernetesManifest class to pass invalidated objects
It would be great if we could pass unvalidated resources.
Since the flag **--validate=false** was added to kubectl. We could probably have a constructor attribute called validate.

### Use Case

I'm trying to add an ALB ingress controller to the EKS cluster. But right now cannot do that through CDK.
I need to add cert-manager which needs to be run without validation.
**Source**:[EKS-ALB-Ingress-Guide](https://docs.aws.amazon.com/eks/latest/userguide/alb-ingress.html)
``` 
kubectl apply --validate=false -f https://github.com/jetstack/cert-manager/releases/download/v1.0.2/cert-manager.yaml
```

### Proposed Solution

Add constructor props to KubernetesManifest class to include **--validate** as an boolean argument.
Also add a validate prop to addManifest() function of @aws-cdk/aws-eks.Cluster class

```
        cluster.add_manifest('alb-ingress-controller-core',*manifest,validate=false)

```
---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by ayush987goyal**
Closes #11763

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by ayush987goyal** [comment]
For kubctl logs:

```
Running command: ['kubectl', 'apply', '--kubeconfig', '/tmp/kubeconfig', '-f', '/tmp/manifest.yaml', '--prune', '-l', 'aws.cdk.eks/prune-c8ae759e1d059c6c67a772adfc4e04d8c732118222', '--validate=false']
```

**Comment by ayush987goyal** [comment]
@iliapolo @eladb Could you please take a look at this change? It has a bug breaking change but I feel this might be okay in a long run since we might have a lot of customizations for the manifests.

**Review by eladb — COMMENTED** [review]
Instead of changing the api for addManifest, would it be sufficient to specify this option at the cluster level (same as prune)?

**Comment by ayush987goyal** [comment]
> Instead of changing the api for addManifest, would it be sufficient to specify this option at the cluster level (same as prune)?

That would set the flag for the complete cluster. What about the case when you only want to skip validation for one particular manifest that you added through `addManifest`?

**Comment by eladb** [comment]
> > Instead of changing the api for addManifest, would it be sufficient to specify this option at the cluster level (same as prune)?
> 
> 
> 
> That would set the flag for the complete cluster. What about the case when you only want to skip validation for one particular manifest that you added through `addManifest`?


Do you think that's a very common use case?

You can still just use `KubernetesManifest` directly.

**Comment by ayush987goyal** [comment]
> You can still just use KubernetesManifest directly.

I am not sure how we can add a manifest intialized via `KubernetesManifest` to the cluster (apart from `addManifest`).

**Comment by eladb** [comment]
> I am not sure how we can add a manifest intialized via `KubernetesManifest` to the cluster (apart from `addManifest`).

See [docs](https://docs.aws.amazon.com/cdk/api/latest/docs/aws-eks-readme.html#kubernetes-manifests)

**Comment by ayush987goyal** [comment]
Oh understood it now (missed that `KubernetesManifest` accespts a cluster!).

I will go ahead and make the changes accordingly to keep it only on `KubernetesManifest`. @eladb Do you think this flag should also be present on the cluster like `prune`? Where I am slightly concerned on adding it to cluster is to reduce the blast radius (caused by potential validation skip) for all the manifests.

**Comment by iliapolo** [comment]
@ayush987goyal I think its ok to only expose `validate: false` on the manifest level and not on the cluster, it stands to reason that this behavior will not normally apply to all manifests in the cluster.

**Comment by ayush987goyal** [comment]
@iliapolo So I am trying to add an integ tests for this by initilializing `KubernetesManifest` by providing the cluster. 

```ts
  private assertManifestWithoutValidation() {
    // apply a kubernetes manifest
    new eks.KubernetesManifest(this, 'HelloAppWithoutValidation', {
      cluster: this.cluster,
      manifest: [{
        apiVersion: 'v1',
        kind: 'Service',
        metadata: { name: 'hello-kubernetes-without-validation' },
        spec: {
          type: 'LoadBalancer',
          ports: [{ port: 80, targetPort: 8080 }],
          selector: { app: 'hello-kubernetes-without-validation' },
        },
      }],
      validate: false,
    });
  }
```

I am getting the following cfn error:

```
Failed to create resource. Error: b'Error from server (AlreadyExists): error when creating "/tmp/manifest.yaml": configmaps "aws-auth" already exists\n' Logs: /aws/lambda/aws-cdk-eks-cluster-test-awscdkaws-Handler886CB40B-4P9SW7KB0BEL at invokeUserFunction (/var/task/framework.js:95:19) at process._tickCallback (internal/process/next_tick.js:68:7)
```

I am thinking if this is related to the scope in which the manifest is created (the context of cluster v/s stack) but not too sure. Do you have any clue on what might be wrong with this?

**Comment by iliapolo** [comment]
@ayush987goyal Yes, we broke our integration tests in this [commit](https://github.com/aws/aws-cdk/pull/12053) :)

@eladb is working on a [fix](https://github.com/aws/aws-cdk/pull/12068). 

You can build an earlier commit to see your logic works, but probably better just to hold off until the fix is merged since you'll need to re-run the tests.

**Comment by ayush987goyal** [comment]
Hi @iliapolo , could you please take a look at this now?

File: packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts:154]: Does the `Service` here serve any purpose? Can this be a simple `ConfigMap`?

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py
Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:27]: Why do we want to default it to `False`?

File: packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts
Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts:154]: Yes it can just be a `ConfigMap`. Will make the change

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:27]: Whoops sorry, good catch! `True` then :)

Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:27]: I think we can leave it as `None` since that will skip the flag altogether and take whatever `kubectl` considers as default.

Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:27]: Yeah I see why you named it `validation_flag` now. The thing is that if we say the default is "whatever kubernetes does when not specifying --validate", than we should advertise that, instead of stating that validation is enabled by default. 

Otherwise we should pass `--validate=true` as the default.

Im also good with changing it to `skipValidation: true` where the default would be `false`, meaning that we let kubernetes do its thing, and then using `None` and `False` here would be exactly the same.

I know it a little pedantic, but its a source of confusion and possible a pretty nasty unexpected behavior in the future.

Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:27]: Yeah I understood the concern now. I think `skipValidation` seems like a better way to handle this. Let me make the change.

File: packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts:152]: Im wondering if we can have an illegal manifest here that would fail if this property isn't satisfied, that would actually test the expected behavior. Maybe just mangle the config map?

Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts:153]: Thanks for this! I was able to notice a couple of bugs because of it:
1. To treat the string-boolean passed from provider correctly ("true"/"false"). Fixed this even for `overwrite`.
2. Include the validation as part of `create` event as well as `update`

Comment (ayush987goyal) [reply packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts:153]: When I say I fixed it even for overwrite, I mean to say that previous implementation would have treated overwrite as truthy since "false" string would have been considered truthy by Python.

File: packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py
Comment (iliapolo) [inline packages/@aws-cdk/aws-eks/lib/kubectl-handler/apply/__init__.py:27]: Lets consolidate all this falsy/truthy shenanigans here, so we can freely use these as booleans later on in the code.

```suggestion
    skip_validation = props.get('SkipValidation', 'false').lower() == 'true'
```

Same for `overwrite ` :)

File: packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts
Comment (iliapolo) [reply packages/@aws-cdk/aws-eks/test/integ.eks-cluster.ts:153]: 👍

**Comment by iliapolo** [comment]
@ayush987goyal Also notice there is a conflict now... :\

**Comment by ayush987goyal** [comment]
@iliapolo Fixed both! :)

**Review by iliapolo — APPROVED** [review]
@ayush987goyal Thanks :)

**Pull request merged**
